### PR TITLE
Add a file lock to the data directory on startup to prevent multiple agents.

### LINKED ISSUE #18243: [Elastic Agent] Prevent a second agent from starting.
Prevent a second instance of the Agent from running to reduce the problems of collision in the configuration or race in the configuration on disk

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Adds an `agent.lock` to the `path.data` directory.

## Why is it important?

Prevents the ability to run multiple agents on the same host.

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [X] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [X] Multiple agents cannot be started at the same time.

## How to test this PR locally

Try to start two `elastic-agent` at the same time on the same host and see that the second one started errors out with `another elastic-agent is already running`.

## Related issues

- Closes #18243

### CODE REVIEW

File: x-pack/elastic-agent/pkg/artifact/download/fs/downloader.go [hunk 1]
```diff
@@ -42,7 +42,7 @@ func NewDownloader(config *artifact.Config) *Downloader {
 func (e *Downloader) Download(_ context.Context, programName, version string) (string, error) {
 	// create a destination directory root/program
 	destinationDir := filepath.Join(e.config.TargetDirectory, programName)
-	if err := os.MkdirAll(destinationDir, os.ModeDir); err != nil {
+	if err := os.MkdirAll(destinationDir, 0755); err != nil {
```
Comment (blakerouse) [inline x-pack/elastic-agent/pkg/artifact/download/fs/downloader.go:45]: This is a drive-by fix on Mac `os.ModeDir` does not create the directory with the proper permissions. `0755` must be used.

File: x-pack/elastic-agent/pkg/agent/cmd/run.go [hunk 2]
```diff
@@ -47,6 +48,12 @@ func run(flags *globalFlags, streams *cli.IOStreams) error {
 		return err
 	}
 
+	locker := application.NewAppLocker(paths.Data())
+	if err := locker.TryLock(); err != nil {
+		return err
+	}
+	defer locker.Unlock()
```
Comment (michalpristas) [inline x-pack/elastic-agent/pkg/agent/cmd/run.go:55]: can you make sure this is handled even if we are killed. defer statements are skipped if SIGINT or SIGTERM are received and it can prevent us from restarting

Comment (blakerouse) [reply x-pack/elastic-agent/pkg/agent/cmd/run.go:55]: I have verified that this does get called in all the cases defined below with signals.

```
signals := make(chan os.Signal, 1)
signal.Notify(signals, syscall.SIGINT, syscall.SIGKILL, syscall.SIGTERM, syscall.SIGQUIT)

<-signals
```

So the defer does get called. I did find a bug in periodic that was preventing `app.Start` from returning to catch the signals. I have fixed that in my most recent commit.
